File: course_1/semester_1/hw8/8_03/List.cpp
#include <iostream>
#include "List.h"
#include "CustomString.h"

using namespace std;

struct ListElement
{
    String *value;
    int count;
    ListElement *next;
};

struct List
{
    ListElement *head;
    int length;
};

List* createList()
{
    List *list = new List;
    list->head = nullptr;
    list->length = 0;
    return list;
}

bool isEmpty(List *l)
{
    return l->length == 0;
}

void deleteList(List *&l)
{
    if(isEmpty(l))
        return;

    ListElement *temp = l->head;

    while (temp != nullptr)
    {
        ListElement *toDelete = temp;
        temp = temp->next;
        deleteString(toDelete->value);
        delete toDelete;
    }

    delete l;
    l = nullptr;
}

ListElement *createListElement(String *value, ListElement *next)
{
    ListElement *newElement = new ListElement;
    newElement->value = value;
    newElement->count = 1;
    newElement->next = next;
    return newElement;
}

int getListLength(List *l)
{
    return l->length;
}

bool addValue(List *l, String *value)
{
    if (isEmpty(l))
    {
        l->head = createListElement(value, nullptr);
        l->length = 1;
        return true;
    }

    ListElement *temp = l->head;

    while (temp->next != nullptr && !equals(value, temp->value))
        temp = temp->next;

    if (equals(value, temp->value))
    {
        temp->count++;
        return false;
    }
    else
    {
        temp->next = createListElement(value, nullptr);
        l->length++;
        return true;
    }
}

void printList(List *l)
{
    if (isEmpty(l))
        return;

    ListElement *temp = l->head;

    while (temp->next != nullptr)
    {
        cout << getChars(temp->value);
        cout << ", ";
        temp = temp->next;
    }

    cout << getChars(temp->value);
}

File: course_1/semester_2/hw1/task_2/src/puzyrev/sem2/hw1/task2/List.java
package puzyrev.sem2.hw1.task2;

public class List<Type> {
    private ListElement head = null;

    public void addValue(Type value) {
        ListElement newElement = new ListElement(value, null);

        if (isEmpty()) {
            head = newElement;
        }

        ListElement temp = head;

        while (temp.next != null && temp.value != value) {
            temp = temp.next;
        }

        if (temp.value != value) {
            temp.next = newElement;
        }
    }

    public boolean deleteValue(Type value) {
        if (isEmpty()) {
            return false;
        }

        if (head.value == value) {
            head = head.next;
            return true;
        }

        ListElement temp = head;

        while (temp.next != null && temp.next.value != value) {
            temp = temp.next;
        }

        if (temp.next == null) {
            return false;
        } else {
            temp.next = temp.next.next;
            return true;
        }
    }

    public boolean isEmpty() {
        return head == null;
    }

    private class ListElement {
        private Type value;
        private ListElement next;

        public ListElement(Type value, ListElement next) {
            this.value = value;
            this.next = next;
        }
    }
}
